Libplanet/Net/TxCompletion.cs

```diff
@@ -0,0 +1,191 @@
+#nullable enable
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Libplanet.Action;
+using Libplanet.Blockchain;
+using Libplanet.Tx;
+using Nito.AsyncEx;
+using Serilog;
+
+namespace Libplanet.Net
+{
+    public class TxCompletion<TPeer, TAction> : IDisposable
+    where TPeer : notnull
+    where TAction : IAction, new()
+    {
+        private readonly ConcurrentDictionary<TPeer, Task> _txSyncTasks;
+        private readonly ConcurrentDictionary<TPeer, ConcurrentBag<TxId>> _requiredTxIds;
+        private readonly CancellationTokenSource _cancellationTokenSource;
+        private readonly BlockChain<TAction> _blockChain;
+        private readonly TxFetcher _txFetcher;
+        private readonly TxBroadcaster _txBroadcaster;
+        private readonly ILogger _logger;
+
+        private bool _disposed;
+
+        public TxCompletion(
+            BlockChain<TAction> blockChain,
+            TxFetcher txFetcher,
+            TxBroadcaster txBroadcaster)
+        {
+            _txSyncTasks = new ConcurrentDictionary<TPeer, Task>();
+            _requiredTxIds = new ConcurrentDictionary<TPeer, ConcurrentBag<TxId>>();
+            _cancellationTokenSource = new CancellationTokenSource();
+            _blockChain = blockChain;
+            _txFetcher = txFetcher;
+            _txBroadcaster = txBroadcaster;
+            TxReceived = new AsyncAutoResetEvent();
+
+            _logger = Log.ForContext<TxCompletion<TPeer, TAction>>();
+        }
+
+        public delegate IAsyncEnumerable<Transaction<TAction>> TxFetcher(
+            TPeer peer,
+            IEnumerable<TxId> txIds,
+            CancellationToken cancellationToken
+        );
+
+        public delegate void TxBroadcaster(TPeer except, IEnumerable<Transaction<TAction>> txs);
+
+        internal AsyncAutoResetEvent TxReceived { get; }
+
+        public async void Dispose()
+        {
+            if (!_disposed)
+            {
+                _cancellationTokenSource.Cancel();
+                var tasks = _txSyncTasks.Values.ToArray();
+                await Task.WhenAll(tasks);
```

Comment:
If we don't have [`IAsyncDisposable`](https://docs.microsoft.com/ko-kr/dotnet/api/system.iasyncdisposable?view=net-5.0) due to .netstandard 2.0 support, it may be better than explicit blocking instead of `async void`.

```suggestion
                Task.WaitAll(tasks);
```